### ISSUE TITLE
Bad allocation thrown on Screen instantiation

### ISSUE THREAD

**Issue by MKMax** [wjakob/nanogui#333]
So I've looked for a good GUI library for some time, and decided to try this one out. I seem to have everything compiled and linked correctly, but when I run any of the examples a bad_alloc exception is thrown when creating a Screen object. A quick look on StackOverflow suggests that it's an issue with a heap allocation on the cpu's part.

I'm using [this](https://github.com/wjakob/nanogui/blob/master/src/example2.cpp) example for the program, and the only things I have added to it was `#define NANOGUI_GLAD` to include the glad header file. 

This seems similar to some of the previous issues regarding Screen's constructor, but I didn't quite find the solution on them.

**Comment by svenevs** [comment]
`#define NANOGUI_GLAD` is meaningless in a single source file, it must be done via CMake: `cmake .. -DNANOGUI_GLAD=ON`.

Internally what this is doing is requesting that GLAD be compiled and statically linked with NanoGUI.

What operating system are you using?  The build system will default to NANOGUI_GLAD for Windows, but shouldn't be necessary on Unix systems.  Where is the `bad_alloc` coming from (compile in debug mode / step through)?  I have a hunch that it's coming from GLFW, but not certain.

**Comment by MKMax** [comment]
Oh yes, I should have mentioned in the OP. I'm using Windows 7 with Visual Studio 2017.

The problem is, no matter how I compile with cmake (following directions on the documentation & including `cmake .. -DNANOGUI_USE_GLAD=ON`) it doesn't seem to include `glad.h` automatically. So I added `#define NANOGUI_GLAD` and that seemed to do the trick for the `opengl.h` file. (I don't see how this could be a problem if I also link against glad, but please correct me if so otherwise).

I also statically linked glad with the project, but that doesn't seem to make a difference.

Here's the bit of code throwing the exception:
```cpp
bool use_gl_4_1 = false;// Set to true to create an OpenGL 4.1 context.
Screen *screen = nullptr;

if (use_gl_4_1) {
    // NanoGUI presents many options for you to utilize at your discretion.
    // See include/nanogui/screen.h for what all of these represent.
    screen = new Screen(Vector2i(500, 700), "NanoGUI test [GL 4.1]",
        /*resizable*/true, /*fullscreen*/false, /*colorBits*/8,
        /*alphaBits*/8, /*depthBits*/24, /*stencilBits*/8,
        /*nSamples*/0, /*glMajor*/4, /*glMinor*/1);
}
else {
    screen = new Screen(Vector2i(500, 700), "NanoGUI test"); // <- bad_alloc here
}
```
It doesn't matter if I use opengl 4.1; it still throws the same exception.
And to answer your question, the exception is thrown when running in debug mode, diagnostic (debugger) mode, and release mode.

**Comment by svenevs** [comment]
Are you working in a parent project, or just compiling from the NanoGUI root repository?  If in a parent project, it sounds like you are missing [`add_definitions(${NANOGUI_EXTRA_DEFS})`](http://nanogui.readthedocs.io/en/latest/compilation.html#required-variables-exposed), which is why you are needing to `#define NANOGUI_GLAD`.

As for the exception, since its coming from the screen constructor you can pretty much assume that it's either coming from GLAD or GLFW.  As a sanity check, does your graphics card support OpenGL 3.3 (very likely unless it is a very old computer)?  3.3 is the default OpenGL context created if not specified explicitly (like in the `use_gl_4_1` part).

The good news is that both GLAD and GLFW are compiled locally, so you should be able to step through with a debugger and see where the crash is coming from.  If you can identify where the crash is coming from, I can provide you with steps to try using a newer version.  Sometimes you get lucky and the bug is already fixed downstream :)

**Comment by svenevs** [comment]
Oh hey that sounds familiar. It's very curious why `glUniform1i` does this indeed. So the issue here is you really need to create a full parent project complete with a `CMakeLists.txt` and git submodule.

To solve the DLL problem for Windows, you need to set `CMAKE_RUNTIME_*_DIRECTORY`. See #310.

Alternatively, build nanogui statically by executing `set(NANOGUI_BUILD_SHARED OFF)` before the `add_subdirectory` call.

Basically, the manual copy of headers and library to a different solution is not necessary, and the real problem here. The segfault may actually come from eigen since you did a 32 bit build but are missing the definitions from nanogui, one of which disables alignment.

I hope that makes sense, I'm in transit and wrote this from a phone :/
